compiler/noirc_evaluator/src/ssa/ssa_gen/context.rs

```diff
@@ -435,6 +448,58 @@ impl<'a> FunctionContext<'a> {
         }
     }
 
+    /// Checks that multiplying two u128 doesn't overflow because both operands are greater or equal than 2^64.
+    /// If both are, then the result is surely greater or equal than 2^128 so it would overflow.
+    /// The operands can still overflow if just one of them is less than 2^64, but in that case the result
+    /// will be less than 2^192 so it fits in a Field value, and acir will check that it fits in a u128.
+    fn check_u128_mul_overflow(&mut self, lhs: ValueId, rhs: ValueId) {
+        let lhs_value = self.builder.current_function.dfg.get_numeric_constant(lhs);
+        let rhs_value = self.builder.current_function.dfg.get_numeric_constant(rhs);
+
+        let two_pow_64 = 1_u128 << 64;
+
+        // If lhs is less than 2^64 then the condition trivially holds.
+        if let Some(value) = lhs_value {
+            if value.to_u128() < two_pow_64 {
+                return;
+            }
+        }
+
+        // Same goes for rhs
+        if let Some(value) = rhs_value {
+            if value.to_u128() < two_pow_64 {
+                return;
+            }
+        }
+
+        let u128 = NumericType::unsigned(128);
+        let two_pow_64 = self.builder.numeric_constant(two_pow_64, u128);
+
+        let res = if lhs_value.is_some() && rhs_value.is_some() {
+            // If both values are known at compile time, at this point we know it overflows
+            self.builder.numeric_constant(FieldElement::one(), u128)
+        } else if lhs_value.is_some() {
+            // If only the left-hand side is known we just need to check that the right-hand side
+            // isn't greater than 2^64
+            self.builder.insert_binary(rhs, BinaryOp::Div, two_pow_64)
+        } else if rhs_value.is_some() {
+            // Same goes for the other side
+            self.builder.insert_binary(lhs, BinaryOp::Div, two_pow_64)
+        } else {
+            // Check both sides
+            let lhs = self.builder.insert_binary(lhs, BinaryOp::Div, two_pow_64);
+            let rhs = self.builder.insert_binary(rhs, BinaryOp::Div, two_pow_64);
+            self.builder.insert_binary(lhs, BinaryOp::Mul { unchecked: true }, rhs)
+        };
+
+        let zero = self.builder.numeric_constant(FieldElement::zero(), u128);
+        self.builder.insert_constrain(
+            res,
+            zero,
+            Some(ConstrainError::StaticString("attempt to multiply with overflow".to_string())),
+        );
```

Comment:
We've also got some type mismatches going on here where we can end up constraining a boolean against a u128.